Source/Android/app/src/main/java/org/dolphinemu/dolphinemu/services/DirectoryInitializationService.java

```diff
@@ -0,0 +1,150 @@
+        {
+            initDolphinDirectories();
+            directoryState = DirectoryInitializationState.DOLPHIN_DIRECTORIES_INITIALIZED;
+            sendBroadcastState(directoryState);
+        }
+        else
+        {
+            sendBroadcastState(DirectoryInitializationState.EXTERNAL_STORAGE_PERMISSION_NEEDED);
+        }
+    }
+
+    private void initDolphinDirectories()
+    {
+        String BaseDir = NativeLibrary.GetUserDirectory();
+        String ConfigDir = BaseDir + File.separator + "Config";
+
+        // Copy assets if needed
+        NativeLibrary.CreateUserFolders();
+        copyAssetFolder("GC", BaseDir + File.separator + "GC", false);
+        copyAssetFolder("Shaders", BaseDir + File.separator + "Shaders", false);
+        copyAssetFolder("Wii", BaseDir + File.separator + "Wii", false);
+
+        // Always copy over the GCPad config in case of change or corruption.
+        // Not a user configurable file.
+        copyAsset("GCPadNew.ini", ConfigDir + File.separator + "GCPadNew.ini", true);
+        copyAsset("WiimoteNew.ini", ConfigDir + File.separator + "WiimoteNew.ini", true);
+    }
+
+    public static boolean areDolphinDirectoriesReady()
+    {
+        return directoryState == DirectoryInitializationState.DOLPHIN_DIRECTORIES_INITIALIZED;
+    }
+
+    private void sendBroadcastState(DirectoryInitializationState state)
+    {
+        Intent localIntent =
+                new Intent(BROADCAST_ACTION)
+                        .putExtra(EXTRA_STATE, state);
+        LocalBroadcastManager.getInstance(this).sendBroadcast(localIntent);
+    }
+
+    private void copyAsset(String asset, String output, Boolean overwrite)
+    {
+        Log.verbose("[DirectoryInitializationService] Copying File " + asset + " to " + output);
+        InputStream in;
+        OutputStream out;
+
+        try
+        {
+            File file = new File(output);
+            if (!file.exists() || overwrite)
+            {
+                in = getAssets().open(asset);
+                out = new FileOutputStream(output);
+                copyFile(in, out);
+                in.close();
+                out.close();
+            }
+        } catch (IOException e)
+        {
+            Log.error("[DirectoryInitializationService] Failed to copy asset file: " + asset + e.getMessage());
+        }
+    }
+
+    private void copyAssetFolder(String assetFolder, String outputFolder, Boolean overwrite)
+    {
+        Log.verbose("[DirectoryInitializationService] Copying Folder " + assetFolder + " to " + outputFolder);
+
+        try
+        {
+            for (String file : getAssets().list(assetFolder))
+            {
+                copyAssetFolder(assetFolder + File.separator + file, outputFolder + File.separator + file, overwrite);
+                copyAsset(assetFolder + File.separator + file, outputFolder + File.separator + file, overwrite);
+            }
+        } catch (IOException e)
+        {
```

Comment:
I'm not sure which of the brace styles we're supposed to use in Dolphin's Java code, but mixing the two like this looks pretty weird. (The end of `copyAsset` has the same problem.)